src/cubic_loader/qlik/ods_qlik.py

```diff
@@ -0,0 +1,402 @@
+            for column in new_columns_to_add:
+                current_schema.append(column)
+            self.update_status(last_schema=current_schema)
+            # re-process failed files with added columns
+            work_files = [
+                (DFMDetails(path=path, ts=re_get_first(path, RE_CDC_TS)), self.etl_status) for path in error_files
+            ]
+            with ThreadPoolExecutor(max_workers=threading_cpu_count()) as pool:
+                for result in pool.map(thread_load_cdc_file, work_files):
+                    dfm_result, new_columns = result
+                    if new_columns is None and dfm_result is not None:
+                        current_cdc_ts = max(current_cdc_ts, dfm_result)
+
+        self.db.vaccuum_analyze(f"{ODS_SCHEMA}.{self.db_history_table}")
+
+        self.update_status(
+            last_cdc_ts=current_cdc_ts,
+            last_schema=current_schema,
+        )
+
+    def rds_fact_table_load(self) -> None:
+        """Load FACT Table records from History Table"""
+        logger = ProcessLogger("load_fact_table", table=self.db_fact_table)
+        schema = self.etl_status.last_schema
+
+        fact_table = f"{ODS_SCHEMA}.{self.db_fact_table}"
+        history_table = f"{ODS_SCHEMA}.{self.db_history_table}"
+
+        table_columns = [col["name"] for col in schema]
+        table_column_str = ",".join(table_columns)
+        key_columns = [col["name"] for col in schema if col["primaryKeyPos"] > 0]
+        key_str = ",".join(key_columns)
+
+        first_vals = []
+        for column in ["header__change_oper"] + table_columns:
+            if column in key_columns:
+                continue
+            q = f"(array_remove(array_agg({column} ORDER BY header__timestamp DESC), NULL))[1] as {column}"
+            first_vals.append(q)
+
+        fact_query = (
+            f"INSERT INTO {fact_table} ({table_column_str})"
+            f" SELECT {table_column_str} FROM"
+            f" ("
+            f" SELECT {key_str}"
+            f" , {','.join(first_vals)}"
+            f" FROM {history_table}"
+            f" GROUP BY {key_str}"
+            f" ) t_load"
+            f" WHERE t_load.header__change_oper <> 'D'"
+            ";"
+        )
+
+        self.db.truncate_table(fact_table)
+        self.db.execute(fact_query)
+
+        self.db.vaccuum_analyze(fact_table)
+
+        logger.log_complete()
+
+    def run_etl(self) -> None:
+        """
+        Run table ETL Process
+
+        Currently this business logic will only process the latest QLIK "Snapshot" that has been issued
+        a feature will need to be added that will handle the issuance of new snapshots that should include:
+            - making sure all outstandind cdc files from last snapshot are loaded
+            - truncating the fact table
+            - resetting the table status file to the new snapshot
+            - re-doing initial load operations to history and fact tables
+        """
```

Comment:
This is the entry point to the ODS-Qlik ETL process. 

Process is clearly laid out in this function.

1. Load snapshot file, if not already loaded. This will only occur on any empty table with no ETL Status file. 
2. Load CDC files from error and archive S3 buckets. Will only load files that have not been processed, as indicated by ETL status file `last_cdc_ts` field. 
3. Load FACT table from _history table of CDC updates. This will load the last non-null value for each column grouped by the table primary key, if the last CDC operation was not a DELETE. 
4. Drop load table used for loading snapshot files.
5. Save status of the ETL S3 status (json) file.